dfget/core/api/download_api.go

```diff
@@ -67,12 +66,71 @@ func (d *downloadAPI) Download(ip string, port int, req *DownloadRequest, timeou
 		}
 	}
 
-	var url string
-	if strings.Contains(req.Path, "://") {
+	var (
+		url      string
+		rangeStr string
+	)
+	if isFromSource(req) {
+		rangeStr = getRealRange(req.PieceRange, headers[config.StrRange])
 		url = req.Path
 	} else {
+		rangeStr = httputils.ConstructRangeStr(req.PieceRange)
 		url = fmt.Sprintf("http://%s:%d%s", ip, port, req.Path)
 	}
+	headers[config.StrRange] = rangeStr
 
 	return httputils.HTTPGetTimeout(url, headers, timeout)
 }
+
+func isFromSource(req *DownloadRequest) bool {
+	return req != nil && strings.Contains(req.Path, "://")
+}
+
+// getRealRange
+// pieceRange: "start-end"
+// rangeHeaderValue: "bytes=sourceStart-sourceEnd"
+// return: "bytes=realStart-realEnd"
+func getRealRange(pieceRange string, rangeHeaderValue string) string {
```

Comment:
How about adding a unit test for this function. 😄

Reply:
done